chat/receipt_handlers.py

```diff
@@ -63,6 +112,76 @@ def __fetch_receipts(messages, user_id):
     return found_receipts
 
 
+def __update_user_conversations(unread_counts, last_read_messages, conn):
+    """
+    Update user conversation table based on un-read counts and
+    last read messages
+    """
+    user_id = current_user_id()
+    schema_name = AsIs(_get_schema_name())
+    for key in unread_counts:
+        delta = unread_counts[key]
+        if delta == 0:
+            continue
+        conn.execute('''
+            UPDATE %(schema_name)s.user_conversation
+            SET
+                "unread_count" = "unread_count" - %(delta)s,
+                "_updated_at" = CURRENT_TIMESTAMP
+            WHERE
+                "conversation" = %(conversation_id)s
+                AND "user" != %(user_id)s
+        ''', {
+            'schema_name': schema_name,
+            'conversation_id': key,
+            'user_id': user_id,
+            'delta': delta
+        })
+        print("updating unread count for user=%s,conversation=%s,delta=%d"
+              % (user_id, key, delta))
+
+    for key in last_read_messages:
+        m = last_read_messages[key]
+        conn.execute('''
+            WITH last_read_message_id AS(
+                SELECT last_read_message AS id FROM
+                %(schema_name)s.user_conversation uc
+                WHERE uc.user = %(user_id)s
+                AND uc.conversation = %(conversation_id)s),
+            message_seq AS (
+                SELECT seq FROM %(schema_name)s.message
+                WHERE _id IN (SELECT id FROM last_read_message_id)
+            )
+            UPDATE %(schema_name)s.user_conversation uc SET
+                last_read_message = CASE WHEN
+                    %(seq)s < (SELECT seq FROM message_seq LIMIT 1)
+                THEN
+                    last_read_message
+                ELSE
+                    %(new_id)s
+                END
+            WHERE uc.user = %(user_id)s
+            AND uc.conversation = %(conversation_id)s
+        ''', {
+            'schema_name': schema_name,
+            'conversation_id': key,
+            'user_id': user_id,
+            'seq': m.record['seq'],
+            'new_id': m.record.id.key
+        })
+        print("updating last_read_message in user=%s,conversation=%s"
+              % (user_id, key))
+
+
+def __update_last_read_messages(last_read_messages, message, key):
+    last_read_message = last_read_messages.get(key, None)
+    if last_read_message is None:
+        last_read_message = message
+    if message.record['seq'] > last_read_message.record['seq']:
+        last_read_message = message
+    last_read_messages[key] = last_read_message
+
+
 def mark_messages(message_ids, mark_delivered, mark_read):
```

Comment:
I am suggesting separate the `upsert receipts` into a separate function, something like `__upsert_receipts `. So the `mark_messages` will become smaller and only represent the sequence of work to be done. Not holding the actual work logic.